src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs

```diff
@@ -0,0 +1,111 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+namespace Microsoft.ServiceFabric.Services.Remoting.V2
+{
+    using System.Diagnostics;
+#if DotNetCoreClr
+    /// <summary>
+    /// Class ActivityIdLogicalCallContext.
+    /// </summary>
+    internal class ActivityIdLogicalCallContext
+    {
+        /// <summary>
+        /// Creates the activity.
+        /// Note : You still need to start it.
+        /// </summary>
+        /// <param name="activityMessage">The activity message.</param>
+        /// <returns>Activity.</returns>
+        internal Activity CreateW3CActivity(string activityMessage = "Create new Activity")
+        {
+            Activity.DefaultIdFormat = ActivityIdFormat.W3C;
+            Activity.ForceDefaultIdFormat = true;
+            var activity = new Activity(activityMessage);
+            return activity;
+        }
+
+        internal Activity StartActivity(IServiceRemotingRequestMessage requestMessage, string activityMessage = "Start new Activity")
+        {
+            var activity = new Activity(activityMessage);
+            string parentId = null;
+            if (requestMessage.GetHeader().ActivityIdParent != null)
+            {
+                Activity.DefaultIdFormat = ActivityIdFormat.W3C;
+                Activity.ForceDefaultIdFormat = true;
+                parentId = requestMessage.GetHeader().ActivityIdParent;
+            }
+            else
+            {
+                parentId = requestMessage.GetHeader().ActivityRequestId;
+            }
+
+            if (!string.IsNullOrEmpty(parentId))
+            {
+                activity.SetParentId(parentId);
+                if (requestMessage.GetHeader().ActivityIdTraceStateHeader != null)
+                {
+                    activity.TraceStateString = requestMessage.GetHeader().ActivityIdTraceStateHeader;
+                }
+            }
+            else
+            {
+                Activity.DefaultIdFormat = ActivityIdFormat.W3C;

```

Comment:
If there is a parent, we dont start the activity?

Reply:
We add the other information, and start it at the end